### PR TITLE
PoC: Standalone Proxy Minion

### PR DESCRIPTION
**Pull request by mirceaulinic**
### What does this PR do?

This is just the very beginning of an idea to add a standalone version of the Proxy Minion. In networking (but not limited to) there are environments that are not very dynamic, or don't need the event driven infra (bad! though I can understand that), and so on. For example, if you only want to push a configuration change once per year you probably don't want a Proxy process running always there.
These are a couple of reasons why we probably should introduce a different flavour of Proxy Minions that are able to control network devices and execute arbitrary commands on demand. Actually there are a couple of approaches:

- An extension to the existing proxy, so we can reuse everything that already exists, and all the proxy modules.
- "Extend the transport-layer abstraction so that it can, as you say, "connect to remote devices over an arbitrary channel" instead of just being something that's designed to be initialized on master/minion startup." @cachedout 
- Provide a completely separate tool that simply invokes the command on the Minions, without having a Master (which may be derived from the first approach, like forcing the command execution without going through the Master).

For the time being I picked the first one, simplifying the Proxy startup by removing the Beacons and the Scheduler (as they don't make sense if you only want a one time command).
The second step was to extend the existing `salt` bin as a `salt-standalone-proxy` that does the same + starts up the lighter Proxy as sub-processes (based on the educated guess to determine what Minions match the target).

The command runs like that (very noisy, will solve that later):

```bash
admin@ip-172-31-9-203:~$ sudo salt-standalone-proxy '*' test.ping
[WARNING ] Temporary logging is already configured
[WARNING ] Temporary logging is already configured
[WARNING ] Console logging already configured
[WARNING ] Console logging already configured
[WARNING ] Logfile logging already configured
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[WARNING ] Logfile logging already configured
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
The Salt ProxyMinion is shutdown.
dummy:
    True
ip-172-31-9-203.us-east-2.compute.internal:
    True
vmx1:
    Minion did not return. [No response]
[WARNING ] ProxyMinion received a SIGTERM. Exiting.
The Salt ProxyMinion is shutdown. ProxyMinion received a SIGTERM. Exited.
```

(`vmx1` was on purpose tested as down, as it couldn't establish the connection to the remote device)

And the process tree is:

```
root     19821  0.0  0.0  40544  3432 pts/0    S+   14:04   0:00  |           \_ sudo salt-standalone-proxy * test.ping
root     19822 16.0  1.0 206100 41704 pts/0    S+   14:04   0:00  |               \_ /usr/bin/python /usr/local/bin/salt-standalone-proxy * test.ping
root     19832 38.0  1.1 452388 47804 pts/0    Sl+  14:04   0:00  |                   \_ /usr/bin/python /usr/local/bin/salt-standalone-proxy * test.ping
root     19833 43.0  1.2 453476 49236 pts/0    Sl+  14:04   0:00  |                   \_ /usr/bin/python /usr/local/bin/salt-standalone-proxy * test.ping
```

I know it's not ideal, but I wanted to discuss first and hear your thoughts.

Pinging @thatch45 @cro @cachedout 

Thanks!

### CODE REVIEW

File: salt/cli/standalone_proxy.py [hunk 1]
```diff
@@ -0,0 +1,448 @@
+# -*- coding: utf-8 -*-
```
Comment (mirceaulinic) [inline salt/cli/standalone_proxy.py]: Maybe it makes more sense to add these to the `salt` cli script...

**Comment by rallytime** [comment]
Hi @mirceaulinic! I think this looks neat. Unfortunately, there is a merge conflict with this PR. Can you rebase and clean that up?

**Comment by mirceaulinic** [comment]
Hi @rallytime - thanks. I will probably rebase this later, I am currently waiting for more feedback and constructive critique.
One of my goals is to be able to run arbitrary commands even without a Master process running.
But I'll be waiting for more input from @thatch45 @cachedout or @cro. Also @gtmanfred I think had an opinion about that.
Thanks!

**Comment by thatch45** [comment]
Sorry about being so slow here, it has been a crazy week as usual and this needs some close attention. But I can say that from my initial look I really like what you are doing here!

**Comment by gtmanfred** [comment]
Yeah, The only thing I was thinking was that it would be super cool to have a `salt-call --local` equivalent for proxy minions, that didn't require access to a master.

Everything else looks like you are on the right track to me!

Thanks!
Daniel

File: salt/cli/standalone_proxy.py (hunk 1 above)
Comment (cachedout) [reply salt/cli/standalone_proxy.py]: This just looks like a copy of the regular minion cli module?

File: salt/minion.py [hunk 2]
```diff
@@ -3469,99 +3472,105 @@ def _post_master_init(self, master):
         self.serial = salt.payload.Serial(self.opts)
         self.mod_opts = self._prep_mod_opts()
         self.matcher = Matcher(self.opts, self.functions)
-        self.beacons = salt.beacons.Beacon(self.opts, self.functions)
+        if self.opts.get('standalone_proxy', False):
+            log.info('Dont need Beacons for this standalone proxy (%s)', self.opts['id'])
```
Comment (cachedout) [inline salt/minion.py]: `Beacons are not required for this standalone proxy`

**Review by cachedout — CHANGES_REQUESTED** [review]

**Comment by cachedout** [comment]
Hi @mirceaulinic. Just getting back to this after a long delay. Apologies!

Where do you want to with this one? Do you want to try to get this into develop and then build on it or did you want to close it and re-open it later on.

**Comment by mirceaulinic** [comment]
Hi @cachedout - thanks for checking in here. The last couple of months have been really insanely busy, but I will hopefully get back to this one and many other soon! Apologies for delaying it.
I don't think this is ready yet to go into develop - I recall there were some things to be sorted beforehand. I understand the maintenance overhead of stale pull requests, so till then please feel free to close and I will reopen later - if that'd work the best for you. No problem from my end at all.

**Comment by rallytime** [comment]
Thanks @mirceaulinic! That sounds good. I'll go ahead and close this and do feel free to reopen when you're ready. :)

**Comment by gtmanfred** [comment]
@mirceaulinic as a follow up to our conversation on slack.  I think this is great, but i think now that `salt-call` can be called using `--proxyid`, that we can do this slightly differently and be more powerful.  And also not have to live inside the salt code base.

I look forward to working with you on this new idea. :)

**Pull request closed without merging**